datafusion/optimizer/src/single_distinct_to_groupby.rs

```diff
@@ -64,22 +67,32 @@ fn is_single_distinct_agg(plan: &LogicalPlan) -> Result<bool> {
     match plan {
         LogicalPlan::Aggregate(Aggregate { aggr_expr, .. }) => {
             let mut fields_set = HashSet::new();
-            let mut distinct_count = 0;
+            let mut aggregate_count = 0;
             for expr in aggr_expr {
                 if let Expr::AggregateFunction(AggregateFunction {
-                    distinct, args, ..
+                    fun,
+                    distinct,
+                    args,
+                    filter,
+                    ..
                 }) = expr
                 {
-                    if *distinct {
-                        distinct_count += 1;
-                    }
-                    for e in args {
-                        fields_set.insert(e.canonical_name());
+                    match filter {
+                        Some(_) => return Ok(false),
+                        None => {
+                            aggregate_count += 1;
+                            if *distinct {
+                                for e in args {
+                                    fields_set.insert(e.canonical_name());
+                                }
+                            } else if !matches!(fun, Sum | Min | Max) {
```

Comment:
Can you please add a comment here explaining why this is checking for only `Sum`, `Min` and `Max`

I think it is because these functions have the property that they can be used to combine themselves -- so like `SUM(SUM(x))` is the same as `SUM(x)`

I think this transformation could be made more general by using a different combination. For example we could support `COUNT` by using `SUM(COUNT(x))`


```sql
SELECT a, COUNT(DINSTINCT b), COUNT(c)
FROM t
GROUP BY a
```

```sql
SELECT a, COUNT(alias1), SUM(alias2) -- <-- This is SUM, not COUNT
FROM (
  SELECT a, b as alias1, COUNT(c) as alias2
  FROM t
  GROUP BY a, b
)
GROUP BY a
```

We can do this as a follow on PR ( I can file a ticket)

Reply:
We could also do crazier stuff for `AVG` like 

```sql
SELECT a, COUNT(DINSTINCT b), AVG(c)
FROM t
GROUP BY a
```

```sql
SELECT a, COUNT(alias1), SUM(alias2) / SUM(alias3) -- <-- This is combining partial sum / counts to compute AVG
FROM (
  SELECT a, b as alias1, SUM(c) as alias2, COUNT(c) as alias3,
  FROM t
  GROUP BY a, b
)
GROUP BY a
```

Reply:
rewrite `Count` requires us to implement `Sum0`. you can refer to the discussion in this [link](https://github.com/apache/arrow-datafusion/pull/8124#issuecomment-1806950843) for details.